DOCSP-9574: Collations

ccho-mongodb opened the pull request:
## Pull Request Info

### Issue JIRA link:
https://jira.mongodb.org/browse/DOCSP-9574

### Docs staging link (requires sign-in on MongoDB Corp SSO):
https://docs-mongodbcom-staging.corp.mongodb.com/d3c6811/java/docsworker-xlarge/DOCSP-9574-collations/fundamentals/collations/

### Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Does it render on staging correctly?
- [x] Are all the links working?
- [x] Are the staging links in the PR description updated?

terakilobyte commented on source/fundamentals/collations.txt:
```diff
@@ -0,0 +1,531 @@
+   :stub-columns: 1
+   :widths: 25,75
+
+   * - Collation Option
+     - Description
```
I wonder if it's worth adding a row for the method name. The names of the options and the names of the method don't appear to follow a consistent scheme, and my first pass at reading this was to read the section intro, look down at the code, then look up at the table.

ccho-mongodb replied:
I've tried adding a column (I think that's what you meant) in another commit, but the output format was difficult to read even after adjusting the column width.  It sounds like that doesn't work for you, so I'd be happy to entertain other suggestions.

Here are some options I've explored in other commits:
- Seealso directives take up too much vertical space which make it difficult to fit the table on a single screen.
- List the method name instead of the collation option: the issue with this is not all the method names are intuitive
- Link to the API documentation from bulleted list of method and omit the description: I think the descriptions I've created are more helpful than those provided in the API reference docs.


I'm not sure what this meant in your comment, could you elaborate and suggest how you might solve the issue?

> my first pass at reading this was to read the section intro, look down at the code, then look up at the table.

terakilobyte replied:
I might just be being too picky, and that's ok. I read the intro, then looked down at the code, referencing the table. The method names don't align with the names of the options as listed, so I was asking if another column could be added. It sounds like that's not ideal to do, so I'm fine leaving it. The content is good, chasing perfect will just slow it down.

ccho-mongodb replied:
Here's an old staged version from a prior commit that shows the additional column https://docs-mongodbcom-staging.corp.mongodb.com/a8ae46a/java/docsworker-xlarge/DOCSP-9574-collations/fundamentals/collations/#collation-options

terakilobyte commented on source/fundamentals/collations.txt:
```diff
@@ -0,0 +1,531 @@
+the :ref:`example collection <collation-example-collection>` and apply
+a collation by specifying the following:
+
+- A group aggregation stage using the ``Aggregates.group() helper to
```
```suggestion
- A group aggregation stage using the ``Aggregates.group()`` helper to
```

terakilobyte reviewed (approved):
Overall looks good, a few things that I noted.

biniona-mongodb reviewed (commented):
Looks great! Learned a ton reading this PR. I mainly commented on sentences that tripped me up.

ccho-mongodb commented on source/fundamentals/collations.txt:
```diff
@@ -0,0 +1,531 @@
+   :end-before: end collationBuilder
+
+For more information on the corresponding methods and parameters they
+take, see the API documentation for :java-docs:`Collation.Builder
```
I don't think we've adopted any standard as both are used across Node and Java documentation. It could be a good topic for a future style discussion though.

biniona-mongodb replied:
Added this as point 18 to the [style doc](https://docs.google.com/document/d/1SxlDvx4MYKBE6-LYGp5i5uuHfY1ZUYYlhc6syEdta5M/edit). I picked a stance on the issue for the sake of discussion, but am open to either (link on just the method name or link on the method name + context).

ccho-mongodb commented on source/fundamentals/collations.txt:
```diff
@@ -0,0 +1,531 @@
+object, set a collation on it, and pass it as a parameter to your call to
+the ``findOneAndUpdate()`` method.
+
+In this example, we demonstrate the following:
```
I think I'm going to stick with my wording here unless you see an issue with my wording.

biniona-mongodb replied:
No worries, totally understand. No issue, just felt suggested change was more conversational. Totally understand if you disagree or prefer your wording as is. 

I am trying out different suggestion types on my first few reviews as well, so feel free to let me know if a suggestion is inappropriate or unhelpful. This was just a tiny nit which was just meant to be a possible version to consider.

ccho-mongodb commented on source/fundamentals/collations.txt:
```diff
@@ -0,0 +1,531 @@
+The following example demonstrates how you can apply a collation when
+retrieving sorted results from the sample collection. To perform this
+operation, call ``find()`` on the example collection and chain the
+``collation()`` and ``sort()`` methods to specify the order in which we want
+to receive the results.
```
Good catch, will fix the "we" -> "you".
The preposition seems extraneous and doesn't seem to provide any value as it doesn't make anything more clear or provide any additional info.

biniona-mongodb replied:
I think that is a fair point on the preposition. I agree that "in" can be removed from the suggested edit.

ccho-mongodb commented on source/fundamentals/collations.txt:
```diff
@@ -0,0 +1,531 @@
+Collation Examples
+------------------
+
+This section contains examples that demonstrate how to use a selection of
```
I wasn't able to find anything incorrect after re-reading the sentence. Could you clarify what you're seeing here that is incorrect?

biniona-mongodb replied:
Oh nothing incorrect, I just thought "some" was a more conversational way of saying "a selection of". Nothing incorrect just a minor style note I had which you can choose to ignore if you do not find it beneficial. 

I wrote "Nit" below the comment to try and signal that this was just a personal opinion, is there a better way I could have signaled that?

ccho-mongodb replied:
Ok, I had seen "Nit", but was just checking whether there was something you thought was incorrect.

ccho-mongodb commented on source/fundamentals/collations.txt:
```diff
@@ -0,0 +1,531 @@
+  { "_id" : 5, "first_name" : "Hannah" }
+
+In the following examples, we specify the "de@collation=phonebook" locale and
+variant collation. This contains rules for prioritizing proper nouns,
```
From re-reading the paragraph, I don't think I've used the pronoun improperly -- if I did, can you find some rule that applies here? 
Somewhat unrelated, I did forgoet to clearly identify what part of "de@collation=phonebook" is the locale and which is the variant which I'll fix.

biniona-mongodb replied:
I don't think it is incorrect, but for whatever reason when I read it my head had trouble understanding that "this" was referencing the collation (my head grouped it with "the following example" and I had to read it twice). I don't think this is a grammar issue, more just me making the sentence easier for myself to understand. Totally understand if you do not want to implement the change, just raising an issue I personally experienced while reading.

ccho-mongodb commented on source/fundamentals/collations.txt:
```diff
@@ -0,0 +1,531 @@
+In the following examples, we specify the "de@collation=phonebook" locale and
+variant collation. This contains rules for prioritizing proper nouns,
+identified by capitalization of the first letter. In an ascending order
+sort, the collation also orders characters with umlauts before the same
```
Next commit updates this paragraph to separate the explanation of locale and variant behavior so this comment probably isn't relevant anymore.

ccho-mongodb commented on source/fundamentals/collations.txt:
```diff
@@ -0,0 +1,531 @@
+
+The following example demonstrates how you can apply a collation when
+retrieving sorted results from the sample collection. To perform this
+operation, call ``find()`` on the example collection and chain the
```
Seems like it could be useful in a future discussion on terminology as "sample" and "example" seem to be used in our documentation.
Since it's within the same paragraph, I'm going to trim the first sentence since it doesn't need to be specific about the collection and can mislead the user into believing the code only works for the example collection.

biniona-mongodb replied:
Added a stance on this issue to the [writing hints and tricks doc](https://docs.google.com/document/d/1SxlDvx4MYKBE6-LYGp5i5uuHfY1ZUYYlhc6syEdta5M/edit) as point 20 with a TODO for further discussion.

ccho-mongodb commented on source/fundamentals/collations.txt:
```diff
@@ -0,0 +1,531 @@
+manual page on :manual:`Use Indexes to Sort Query Results </tutorial/sort-results-with-indexes/>`.
+
+The following code snippet shows an example query operation that specifies
+the Icelandic ("is") collation that is not covered by the default collation
```
I agree the length can make the sentence more complex. Since the sentence is describing a single item, I think dividing it makes the sentences too short which can be misinterpreted as being talked down to. E.g. "The brown dog that lives with its owner ran to the store." vs "The brown dog lives with its owner. The brown dog ran to the store." Instead, I'll present multiple (including information that wasn't presented explicitly before) conditions which can be listed in fragments.

biniona-mongodb replied:
I like the edits! Solves the issue I experienced and avoids a potentially patronizing tone. Resolving conversation.

ccho-mongodb commented on source/fundamentals/collations.txt:
```diff
@@ -0,0 +1,531 @@
+aggregation pipeline. To perform an aggregation, call the ``aggregate()``
+method on a ``MongoCollection`` object.
+
+To specify a collation for an aggregate operation, call the ``collation()``
```
Good catch, I meant aggregation operation here.

ccho-mongodb commented on source/fundamentals/collations.txt:
```diff
@@ -0,0 +1,531 @@
+
+   You can assign a default collation to a collection only when you create
+   that collection. To use collations for supported operations on an
+   existing collection, create an index with the collation and specify the
+   same collation in your operations on it.
```
Thanks for catching this. I'll rewrite, I think I missed some other details as well.

ccho-mongodb commented on source/fundamentals/collations.txt:
```diff
@@ -0,0 +1,531 @@
+       | :java-docs:`collationStrength() <apidocs/mongodb-driver-core/com/mongodb/client/model/Collation.Builder.html#collationStrength(com.mongodb.client.model.CollationStrength)>` API documentation
+
+   * - Locale
+     - | **Required**. The ICU locale code for language and variant.
```
Not sure as there isn't a specific ordering here. I'll move it to the top as that shouldn't cause any issues.

ccho-mongodb replied:
Upon examination, I realized they are ordered by method name. I'll move it to the top anyway since it's probably the most important one.

biniona-mongodb replied:
I made a note in the [writing hints and tricks doc](url) as point 21 specifying that required parameters should go first with the caveat that it requires further discussions.

biniona-mongodb commented on source/fundamentals/collations.txt:
```diff
@@ -0,0 +1,539 @@
+--------
+
+This guide shows you how to use **collations** with MongoDB to order your
+query or aggregate results by string values. A collation is a set of
```
```suggestion
query or aggregation results by string values. A collation is a set of
```

Aggregate vs Aggregation terminology. If you think aggregate is more appropriate here totally understand, but seems related to https://github.com/mongodb/docs-java/pull/70#discussion_r605236948

ccho-mongodb replied:
Good catch, I'll update.

biniona-mongodb reviewed (approved):
LGTM! One possible thing to consider related to a previous comment. Looks like a really helpful page.

jyemin commented on source/fundamentals/collations.txt:
```diff
@@ -0,0 +1,539 @@
+   "<locale code>@collation=<variant code>"
+
+Specify a valid locale in the ``locale code`` placeholder and optionally a
+variant in the ``variant code`` placeholder. The following example specifies
```
I'm unclear how to parse this sentence.  What does `placeholder` mean in this context?

ccho-mongodb replied:
The placeholder is the presentation of the text between the angled braces. In this paragraph, "<locale code>" is a placeholder for locale code. Since it doesn't seem as intuitive as I assumed, what would you suggest to make this clear to the reader?

jyemin replied:
I read it again and now it makes sense.  I just didn't read `"<locale code>@collation=<variant code>"` very carefully and didn't even notice the placeholders

ccho-mongodb replied:
I'll remove that sentence since it repeats the sentence prior to the example. Hopefully, that helps guide the reader directly into the examples through which they can infer the format.

jyemin commented on source/fundamentals/collations.txt:
```diff
@@ -0,0 +1,539 @@
+
+.. note::
+
+   You can only assign a default collation to a collection when you
```
Nit: It's a little weird reading this to have a note like this that refers to the very next section on the very same page.

ccho-mongodb replied:
Ok, I'll move it to a spot prior to the example. Let me know if that looks better to you, once committed.

jyemin commented on source/fundamentals/collations.txt:
````````````````````diff
@@ -0,0 +1,539 @@
+```````````````````
+
+This section demonstrates how you can specify a collation in an aggregation
+operation. In an aggregation operation, you can specify a series
````````````````````
`series aggregation stages` => `series of aggregation stages`

jyemin commented on source/fundamentals/collations.txt:
```diff
@@ -0,0 +1,539 @@
+.. code-block:: json
+   :copyable: false
+
+   {"_id": "Günter", "nameCount": 1}
```
The example would be more compelling if one of the groups had more than one member.  Maybe an example that uses a collation that would collapse both `Gunter`s into a single group?  Is that possible?

ccho-mongodb replied:
I'll try to find a different collation that might ignore umlauts.

ccho-mongodb commented:
Hi @jyemin, thanks for the review. I've addressed all your comments and updated the preview url in the PR description. I wasn't sure if you had a suggestion for this comment or whether it was just a question https://github.com/mongodb/docs-java/pull/70#discussion_r606312845

jyemin commented on source/includes/fundamentals/code-snippets/CollationCollectionExample.java:
```diff
@@ -190,12 +193,13 @@ private static void operationCollation(MongoCollection collection) {
     public static void main(String[] args) {
 
-        String uri = "<MongoDB connection URI>";
+        String uri = "mongodb://localhost:27017";
```
Did you mean to leave this change in?

ccho-mongodb replied:
No, thanks for catching this. I wonder if maybe the checked-in class shouldn't even contain a main method since it's not included in any of the code snippets.

jyemin commented on source/includes/fundamentals/code-snippets/CollationCollectionExample.java:
```diff
@@ -190,12 +193,13 @@ private static void operationCollation(MongoCollection collection) {
 
             MongoDatabase database = mongoClient.getDatabase("fundamentals_example");
-//            MongoCollection<Document> collection = database.getCollection("phonebook");
+            // MongoCollection<Document> collection = database.getCollection("phonebook");
```
Did you mean to include this change?

ccho-mongodb replied:
Nope, will fix

jyemin reviewed (approved):
Couple of small questions, but no need for another review.  Fix if necessary.
